spec/ndt7-protocol.md

```diff
@@ -192,7 +192,110 @@ implementations will likely round bigger (or smaller) numbers to the nearest
 make sure that it does not mit values that a `int53` cannot represent. The
 proper action to take in this case is currently unspecified.
 
-# Reference implementation
+## Server discovery
+
+This section explains how a client can get the FQDN of one or more
+ndt7-enabled servers for the purpose of running tests. Of course, this
+section only applies to clients using M-Lab infrastructure. Clients
+using other server infrastructure MUST refer to the documentation
+provided by such infrastructure instead.
+
+Clients:
+
+1. SHOULD use [mlab-ns](https://github.com/m-lab/mlab-ns);
+
+2. MUST query for the `ndt_ssl` mlab-ns tool (see below for a
+description of how a real request would look like);
+
+3. MUST set `User-Agent` to identify themselves;
+
+4. SHOULD use the `policy=geo_options` policy that causes mlab-ns to
+return a set of nearby servers;
+
+5. if the connection to the first returned server is down, they SHOULD
+try with subsequent servers, and fail the test if all the returned
+servers appear to be down;
+
+6. MUST handle gracefully the case where M-Lab is out of capacity, which
+is identifiable by mlab-ns returning `200` along with an empty set of servers;
+
+7. MUST use the following retry policy (if applicable): extract a random
+number of seconds from an exponential distribution with average equal
+to `3,600` seconds, and wait that number of seconds before trying again;
+
+8. MUST (of course) interpret `200` as success and any other status
+code as failure, and act accordingly.
+
+For the purpose of the beta testing phase of `ndt7`, and whenever they wish
+to use non production infrastructure, client MUST use the following base
+URL for mlab-ns:
+
+```
+https://locate-dot-mlab-staging.appspot.com
+```
+
+Otherwise, they MUST use the following base URL (which will not work
+until ndt7 is deployed on the production infrastructure):
+
+```
+https://mlab-ns.appspot.com
+```
+
+Thus, the full URL to contact mlab-ns MUST be:
+
+```
+${baseURL}/ndt_ssl?policy=geo_options
```

Comment:
Why these url params? They could also just ask for a single server using the bare URL.

Reply:
Because having more than one server is more robust: a server may be down and we still want to run a test with a backup server. Isn't this the reason why `geo_policy` was added and is used by specific clients to have more robustness?

Reply:
Yes, but that was back when our systems were less reliable and mlab-ns had a 45 minute update latency.  Now it has a much tighter (under 2 minutes) freshness bound, and is serving a much more accurate view of the world. There's a lot less need for the client to need to get a list of servers unless they have very high reliability needs.

Reply:
Awesome: I didn't know about this and really like to avoid to loop.